### PR TITLE
Ticket3939 interpolation accuracy

### PR DESCRIPTION
**Pull request by LiamPanchaud**
Includes spline interpolation developer defined function plus googletests to ensure correct working.

### CODE REVIEW

File: Keithley_2700Sup/user1DTableSub.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        double x:           value to be evaluated by fit
+        void ** dpvt:       pointer to record's dpvt field for storing data between calls       
+*/
+static double user1DTableSub(bool isInit, double x_data[], double y_data[], int len_arr, double x, void ** dpvt) {
+
+    gsl_interp_accel *acc;
+    gsl_spline *spline;
+    struct Spline_data_container *spline_d_cont;
+
+    if (isInit) {
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/user1DTableSub.c:57]: How expensive is the fitting process, and how often will the spline need to be evaluated? If it is quick and cheap, or isn't evaluated very often, I would consider dropping the Init flag and have it always fit the spline

File: Keithley_2700Sup/user1DTableSub.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+			y_data[i] = y_data[j];
+			y_data[j] = value_buffer;
+			j--;
+		}
+
+        spline_d_cont = (struct Spline_data_container *)malloc(sizeof(struct Spline_data_container));
+
+        acc = gsl_interp_accel_alloc();
+        spline = gsl_spline_alloc(gsl_interp_cspline, len_arr);
+        gsl_spline_init(spline, x_data, y_data, len_arr);
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/user1DTableSub.c:73]: The GSL documentation example shows these quantities getting freed after their use: 
https://www.gnu.org/software/gsl/doc/html/interp.html#d-interpolation-example-programs

Is this getting performed? If not, could this be a memory leak?

File: run_tests.bat [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+:: Move XML file to test-reports folder
+
+if not exist %TOP%\test-reports (
+    mkdir %TOP%\test-reports
+)
+
+move "TEST-*.xml" "%TOP%\test-reports\"
+
+exit /B %keithley_2700_tests_failed%
```
Comment (Alistair-McGann-Tessella) [inline run_tests.bat]: Needs a newline

File: Keithley_2700Sup/drift_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    if ((!isnan(temp_change_over_time)) && (isnan(previous_drift))) {
+        new_drift = temp_change_over_time;
+        printf("Assigned new drift with nans to %f\n", new_drift);
+    }
+    else {
+        new_drift = temp_change_over_time + previous_drift;
+        printf("Assigned new drift without nans to %f\n", new_drift);
+    }
+    return new_drift;
+}
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/drift_utils.cpp]: Needs a newline

File: Keithley_2700Sup/drift_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    // If temp_change_over_time IS a number AND the previous drift is NOT a number
+    // (i.e. this is the first reading) set the drift to equal only the new 
+    // temp_change_over_time, else it is the new temp_change_over_time summed with
+    // the previously calculated drift
+    if ((!isnan(temp_change_over_time)) && (isnan(previous_drift))) {
+        new_drift = temp_change_over_time;
+        printf("Assigned new drift with nans to %f\n", new_drift);
+    }
+    else {
+        new_drift = temp_change_over_time + previous_drift;
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/drift_utils.cpp:31]: This looks different to the algorithm on the wiki

File: Keithley_2700Sup/drift_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+    temp_change_over_time = temp_change_over_time/SMOOTHING_FACTOR;
+
+    // If temp_change_over_time IS a number AND the previous drift is NOT a number
+    // (i.e. this is the first reading) set the drift to equal only the new 
+    // temp_change_over_time, else it is the new temp_change_over_time summed with
+    // the previously calculated drift
+    if ((!isnan(temp_change_over_time)) && (isnan(previous_drift))) {
+        new_drift = temp_change_over_time;
+        printf("Assigned new drift with nans to %f\n", new_drift);
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/drift_utils.cpp]: How often will this write to the IOC console? Is this a required statement for logging?

File: Keithley_2700Sup/parse_buffer_readings.c [hunk 7]
```diff
@@ -24,30 +24,30 @@ static void assign_value_to_pv(double *val, epicsEnum16 ftv, double reading) {
...
 	* 	a[i+1] = channel timestamp
 	* 	a[i+2] = channel 
 	*/
-	
+
 	double *a;
 	// We have 3 values per channel - reading, timestamp and channel
 	const int channel_values_count = 3;
-	long i = 0;
+	unsigned long i = 0;
```
Comment (Alistair-McGann-Tessella) [inline Keithley_2700Sup/parse_buffer_readings.c]: I recently got feedback that use of unsigned types is discouraged by the Google C++ style guidelines. I don't see the use here is a problem, but I am curious as to why you made this change.

**Review by Alistair-McGann-Tessella — CHANGES_REQUESTED** [review]

File: Keithley_2700Sup/user1DTableSub.c (hunk 1 above)
Comment (FreddieAkeroyd) [reply Keithley_2700Sup/user1DTableSub.c:57]: You mean drop the Init flag from the implementation but not the interface?

File: Keithley_2700Sup/parse_buffer_readings.c (hunk 7 above)
Comment (LiamPanchaud) [reply Keithley_2700Sup/parse_buffer_readings.c]: I _think_ it was to mitigate a compiler warning where something unsighed was being compared to `i` (which was signed at the time). Some field in the record (I suspect it was `noa`) uses an unsigned value, and I thought it best to match the types

File: Keithley_2700Sup/drift_utils.cpp (hunk 6 above)
Comment (LiamPanchaud) [reply Keithley_2700Sup/drift_utils.cpp]: Removed

File: Keithley_2700Sup/user1DTableSub.c (hunk 1 above)
Comment (LiamPanchaud) [reply Keithley_2700Sup/user1DTableSub.c:57]: It's _reasonably_ expensive, and although I have no doubt that the computer could handle doing this, it doesn't seem like it's worth making it do all the extra work of fitting every few seconds

File: Keithley_2700Sup/user1DTableSub.c (hunk 2 above)
Comment (LiamPanchaud) [reply Keithley_2700Sup/user1DTableSub.c:73]: This is now noted in the [convert record wiki page](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Convert-Record#tips-tricks-and-gotchas) in point 3

**Review by Alistair-McGann-Tessella — APPROVED** [review]

**Pull request merged**